plutus-tx-plugin/test/Plugin/Data/recursive/interListConstruct.tplc.golden

```diff
@@ -0,0 +1,749 @@
+                              )
+                            )
+                            (lam
+                              dat_i0
+                              (fun (type) (fun (type) (type)))
+                              [ [ dat_i1 a_i3 ] b_i2 ]
+                            )
+                          )
+                        )
+                      )
+                      b_i4
+                    ]
+                    a_i5
+                  ]
+                  (iwrap
+                    (lam
+                      rec_i0
+                      (fun (fun (fun (type) (fun (type) (type))) (type)) (type))
+                      (lam
+                        spine_i0
+                        (fun (fun (type) (fun (type) (type))) (type))
+                        [
+                          spine_i1
+                          [
+                            (lam
+                              InterList_i0
+                              (fun (type) (fun (type) (type)))
+                              (lam
+                                a_i0
+                                (type)
+                                (lam
+                                  b_i0
+                                  (type)
+                                  (sop
+                                    [a_i2 b_i1 [ [ InterList_i3 b_i1 ] a_i2 ]]
+                                    []
+                                  )
+                                )
+                              )
+                            )
+                            (lam
+                              a_i0
+                              (type)
+                              (lam
+                                b_i0
+                                (type)
+                                [
+                                  rec_i4
+                                  (lam
+                                    dat_i0
+                                    (fun (type) (fun (type) (type)))
+                                    [ [ dat_i1 a_i3 ] b_i2 ]
+                                  )
+                                ]
+                              )
+                            )
+                          ]
+                        ]
+                      )
+                    )
+                    (lam
+                      dat_i0
+                      (fun (type) (fun (type) (type)))
+                      [ [ dat_i1 a_i6 ] b_i5 ]
+                    )
+                    (constr
+                      [
+                        (lam
+                          InterList_i0
+                          (fun (type) (fun (type) (type)))
+                          (sop [a_i6 b_i5 [ [ InterList_i1 b_i5 ] a_i6 ]] [])
+                        )
+                        (lam
+                          a_i0
+                          (type)
+                          (lam
+                            b_i0
+                            (type)
+                            (ifix
```

Comment:
@michaelpj this is what the line you pointed me to produces for the `InterList` example. It's indeed what one'd expect and it doesn't look problematic (i.e. I take back my suspicion that there can be something wrong with it). I do still find it unsettling. How come we have to inline the entire definition of the data type whenever we define a constructor of that data type? Maybe we're missing some interleaving opportunity or I don't know?

Anyways, I'm going to keep thinking about it, but probably after I find a good way of rendering this test using the readable syntax, as I'm not particularly keen on studying 749 lines of S-expression abomination.

Reply:
I think I convinced myself that it essentially corresponds to the typing rule for `iwrap`, where we do exactly substitute the whole definition of the recursive type for itself.

Reply:
I _think_, it indeed plays out like that, but via a different pathway, kinda. With `iwrap` you substitute both the data type _and_ the index -- that's the whole idea behind `ifix`. Here you only substitute the type in the original pattern functor body with `n` type variables, which is very much not was intended with the whole data type compilation machinery... But it works! And it does kinda make sense. My brain hurts.

I think I have an idea on how to reduce the duplication. I'll create a ticket and return to it once I'm done with BLS builtins.